src/librustc_typeck/astconv.rs

```diff
@@ -648,7 +656,87 @@ impl<'o, 'tcx> dyn AstConv<'tcx> + 'o {
                                 if arg_count.correct.is_ok()
                                     && arg_count.explicit_late_bound == ExplicitLateBound::No
                                 {
-                                    Self::generic_arg_mismatch_err(tcx.sess, arg, kind.descr());
+                                    // We're going to iterate over the parameters to sort them out, and
+                                    // show that order to the user as a possible order for the parameters
+                                    let mut ordered_params = defs.params.clone();
+                                    ordered_params.sort_by_key(|param| match param.kind {
+                                        GenericParamDefKind::Lifetime => ParamKindOrd::Lifetime,
+                                        GenericParamDefKind::Type { .. } => ParamKindOrd::Type,
+                                        GenericParamDefKind::Const => ParamKindOrd::Const,
+                                    });
+
+                                    let is_lifetime_present = ordered_params
```

Comment:
Rather than iterating through `ordered_params` three times, let's instead iterate through once (in a `for` loop) and set these variables. You could use a `HashSet<ParamKindOrd>`, which might be a little cleaner.